docs/man/XLine.3
.TH "XLine" 3 "30 Apr 2004" "InspIRCd" \" -*- nroff -*-
.ad l
.nh
.SH NAME
XLine \- XLine is the base class for ban lines such as G lines and K lines. 

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <xline.h>\fP
.PP
Inherits \fBclassbase\fP.
.PP
Inherited by \fBGLine\fP, \fBKLine\fP, \fBQLine\fP, and \fBZLine\fP.
.PP
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "time_t \fBset_time\fP"
.br
.RI "\fIThe time the line was added.\fP"
.ti -1c
.RI "long \fBduration\fP"
.br
.RI "\fIThe duration of the ban, or 0 if permenant.\fP"
.ti -1c
.RI "char \fBsource\fP [MAXBUF]"
.br
.RI "\fISource of the ban.\fP"
.ti -1c
.RI "char \fBreason\fP [MAXBUF]"
.br
.RI "\fIReason for the ban.\fP"
.ti -1c
.RI "long \fBn_matches\fP"
.br
.RI "\fINumber of times the core matches the ban, for statistics.\fP"
.in -1c
.SH "Detailed Description"
.PP 
XLine is the base class for ban lines such as G lines and K lines.Definition at line 18 of file xline.h.
.SH "Member Data Documentation"
.PP 
.SS "long XLine::duration"
.PP
The duration of the ban, or 0 if permenant.Definition at line 28 of file xline.h.
.SS "long XLine::n_matches"
.PP
Number of times the core matches the ban, for statistics.Definition at line 40 of file xline.h.
.SS "char XLine::reason[MAXBUF]"
.PP
Reason for the ban.Definition at line 36 of file xline.h.
.SS "time_t XLine::set_time"
.PP
The time the line was added.Definition at line 24 of file xline.h.
.SS "char XLine::source[MAXBUF]"
.PP
Source of the ban.This can be a servername or an oper nicknameDefinition at line 32 of file xline.h.

.SH "Author"
.PP 
Generated automatically by Doxygen for InspIRCd from the source code.


docs/ChangeLog
1.0 Final

    * Optimized WriteServ - why use %s!%s@%s when we have the more optimal userrec::GetFullHost() ?
    * Added new event into m_spanningtree, event id send_metadata, which allows arbitary posting of metadata outside of the netburst, see line 529 of m_ssl_gnutls for an example of how to use it.
    * Added check that forbids configuration as root >:-)
    * Major updates to CBANs...like they're reasonably functional
    * Added support for TYPE_OTHER metadata, where void* target will be NULL, e.g. its not bound to a user or a nick
    * Fix to send 403 on invalid channel name
    * Moved Write() call so that it can send the NICK message properly and force the client to change
    * Allow ulines to still notice the channel regardless of +T
    * Created OnPostOper that can be gauranteed to be called *after* opering (by ALL modules) is done
    * "Oper of unknown type" -> "313 %s %s :is an IRC operator"
    * Auto-assume "+" on the start of opermode block if not included
    * Removed example commands that never actually existed
    * OpenSSL detection and cert generation in ./configure
    * Changed length of password field to 64, giving us a maximum password length of 63 plus null terminator
    * Fixed uninitialized ExtMode::list
    * Added some extra debug
    * Make m_services use InsertMode (are we done yet?)
    * Changed On005Numeric to use InsertMode (we left one out?)
    * Deleted the semicolon that had half the dev team wondering...
    * Incorporated Craig's friendlier SQL class (documentation to come later)
    * m_sqlauth: Fix small memory leaks in some cases (bad password, bad query, lost mysql server connection etc..)
    * Correctly initialize userrec::password (should fix a few m_sql issues)
    * Add m_conn_waitpong, require clients to reply with a correct PONG to a server PING before they can connect
    * Remove unused includes and rename classes from ModuleSSL to ModuleSSLGnuTLS so we can be consistent across multiple SSL modules
    * Added warnings to opers when session limits are exceeded
    * Improved IP handling. Now uses in_addr to store client ip, not char[16]!, Added global and local session limits
    * Problems removing +t due to wrong bit-value being used in comparison
    * Altered so that if running in -debug mode, will not erase its .so files from /tmp to allow a clean backtrace
    * Removed unneccessary strlcpy
    * Removed a load of unneeded includes (poll.h etc)
    * Added <ident:timeout> example so that people remember to set this when using m_ident.so
    * Default ident timeout to 1 sec if not defined (a bit low, but better than 0!)
    * Updated m_ident to use inspsocket, should be more reliable
    * Added missing modules (newer ones)
    * Removed excessive debug for something thats now fixed
    * Merged Andy Church's isnick() and isident() patches - faster, more efficient, and doesnt accidentally allow "-" as the first character of a nickname, Also adapted to iterate over a char* pointer rather than using unsigned int for the for() loops
    * Added m_opersha256 module created by Special
    * Add module for anti-autorejoin channel mode +J
    * Changes to INVEX and EXCEPTS 005 tokens
    * Changed +l to drop the mode if the limit is same as current., Optimized check on numeric chars
    * Remove last relic of last goto in inspircd (an unused label)
    * Fixed to not have ohv in CHANMODES type A,, Added proper MAXMODES output in 005 instead of hard coded '13'
    * One goto is now no gotos.
    * Change m_redirect, m_joinflood and m_messageflood to put their modes in the correct section and to use InsertMode()
    * Add InsertMode function to helperfuncs.(cpp|h) for easily adding modes to the CHANMODES section of the 005 numeric
    * Added patch from Andy Church re spaces in SETNAME
    * Fixed line endings when skipping ssl cert generation
    * Documented some functions that were missing comments
    * Properly check that existing key resides in conf/ and not current dir
    * Added PRIORITY_BEFORE and PRIORITY_AFTER (see src/modules/m_hostchange.cpp for how it works, function Prioritize())
    * Added -modupdate to regenerate modules makefile, Added SSL option, as well as a 'Build Certificate' section (If avaliable)
    * Adding the +e and +I modules m_banexception and m_inviteexception, also adding their shared utility header
    * Added trick to allow local server to unset -r only in one specific condition (the nickchange)
    * Added joinflood +j
    * Made bans check both real and fake host
    * Patched to allow FTOPIC from users (so that chanserv can update the settime and the topic still come from chanserv)
    * Added <link:hidden> to docs
    * Optional IP-Hiding for servers
    * Use libgnutls-config --libs to configure linker options
    * Make relative paths in the config file tags be relative to the insp config directory
    * Change 974 numeric to 490 to avoid collision with Insp's failed to load module error
    * Make m_sslmodes messages appear in the channel window
    * Updated with new docs for maxtargets
    * Updated 005 to show changeable MAXTARGETS
    * Allowed for configuration of maxtargets in <options:maxtargets>
    * Wrong nicklen is shown in the 005 numeric, fixed (noticed by ThaPrince)
    * Initial commit of m_sslmodes, supplying chanmode +z
    * Allow for <bind> tags with no or empty type, as well as 'clients' ones
    * Set vendor and static module flags...even though the booms from trying to unload it would be fun sad.gif
    * Make errors loading certificates DEFAULT level
    * Inital commit of the first SSL module \o/ \o/ \o/ Also Ommeh's first official svn commit \o/
    * Added MakeWildHost to userrec (*!*@dhost), Added ban stuff to messageflood
    * Fixes to server kick (i hope)
    * And yae, the compiler sayeth to brain, fix thine warning. and thou warning art fixeth.
    * Made '*' ban-specifier work

servicesd/src/xline.cpp
/* XLine functions.
 *
 * (C) 2003-2023 Anope Team
 * Contact us at team@anope.org
 *
 * Please read COPYING and README for further details.
 *
 * Based on the original code of Epona by Lara.
 * Based on the original code of Services by Andy Church.
 */

#include "services.h"
#include "modules.h"
#include "xline.h"
#include "users.h"
#include "sockets.h"
#include "regexpr.h"
#include "config.h"
#include "commands.h"
#include "servers.h"

/* List of XLine managers we check users against in XLineManager::CheckAll */
std::list<XLineManager *> XLineManager::XLineManagers;
Serialize::Checker<std::multimap<Anope::string, XLine *, ci::less> >
XLineManager::XLinesByUID("XLine");

void XLine::Init() {
    if (this->mask.length() >= 2 && this->mask[0] == '/'
            && this->mask[this->mask.length() - 1] == '/'
            && !Config->GetBlock("options")->Get<const Anope::string>("regexengine").empty()) {
        Anope::string stripped_mask = this->mask.substr(1, this->mask.length() - 2);

        ServiceReference<RegexProvider> provider("Regex",
                Config->GetBlock("options")->Get<const Anope::string>("regexengine"));
        if (provider) {
            try {
                this->regex = provider->Compile(stripped_mask);
            } catch (const RegexException &ex) {
                Log(LOG_DEBUG) << ex.GetReason();
            }
        }
    }

    size_t nick_t = this->mask.find('!');
    if (nick_t != Anope::string::npos) {
        nick = this->mask.substr(0, nick_t);
    }

    size_t user_t = this->mask.find('!'), host_t = this->mask.find('@');
    if (host_t != Anope::string::npos) {
        if (user_t != Anope::string::npos && host_t > user_t) {
            user = this->mask.substr(user_t + 1, host_t - user_t - 1);
        } else {
            user = this->mask.substr(0, host_t);
        }
    }

    size_t real_t = this->mask.find('#');
    if (host_t != Anope::string::npos) {
        if (real_t != Anope::string::npos && real_t > host_t) {
            host = this->mask.substr(host_t + 1, real_t - host_t - 1);
        } else {
            host = this->mask.substr(host_t + 1);
        }
    } else {
        if (real_t != Anope::string::npos) {
            host = this->mask.substr(0, real_t);
        } else {
            host = this->mask;
        }
    }

    if (real_t != Anope::string::npos) {
        real = this->mask.substr(real_t + 1);
    }

    if (host.find('/') != Anope::string::npos) {
        c = new cidr(host);
        if (!c->valid()) {
            delete c;
            c = NULL;
        }
    }
}

XLine::XLine(const Anope::string &ma, const Anope::string &r,
             const Anope::string &uid) : Serializable("XLine"), mask(ma), by(Me->GetName()),
    created(0), expires(0), reason(r), id(uid) {
    regex = NULL;
    manager = NULL;
    c = NULL;

    this->Init();
}

XLine::XLine(const Anope::string &ma, const Anope::string &b, const time_t ex,
             const Anope::string &r, const Anope::string &uid) : Serializable("XLine"),
    mask(ma), by(b), created(Anope::CurTime), expires(ex), reason(r), id(uid) {
    regex = NULL;
    manager = NULL;
    c = NULL;

    this->Init();
}

XLine::~XLine() {
    if (manager) {
        manager->RemoveXLine(this);
    }

    delete regex;
    delete c;
}

const Anope::string &XLine::GetNick() const {
    return nick;
}

const Anope::string &XLine::GetUser() const {
    return user;
}

const Anope::string &XLine::GetHost() const {
    return host;
}

const Anope::string &XLine::GetReal() const {
    return real;
}

Anope::string XLine::GetReason() const {
    Anope::string r = this->reason;
    if (!this->id.empty()) {
        r += " (ID: " + this->id + ")";
    }
    return r;
}

bool XLine::HasNickOrReal() const {
    return !this->GetNick().empty() || !this->GetReal().empty();
}

bool XLine::IsRegex() const {
    return !this->mask.empty() && this->mask[0] == '/'
           && this->mask[this->mask.length() - 1] == '/';
}

void XLine::Serialize(Serialize::Data &data) const {
    data["mask"] << this->mask;
    data["by"] << this->by;
    data["created"] << this->created;
    data["expires"] << this->expires;
    data["reason"] << this->reason;
    data["uid"] << this->id;
    if (this->manager) {
        data["manager"] << this->manager->name;
    }
}

Serializable* XLine::Unserialize(Serializable *obj, Serialize::Data &data) {
    Anope::string smanager;

    data["manager"] >> smanager;

    ServiceReference<XLineManager> xlm("XLineManager", smanager);
    if (!xlm) {
        return NULL;
    }

    XLine *xl;
    if (obj) {
        xl = anope_dynamic_static_cast<XLine *>(obj);
        data["mask"] >> xl->mask;
        data["by"] >> xl->by;
        data["reason"] >> xl->reason;
        data["uid"] >> xl->id;

        if (xlm != xl->manager) {
            xl->manager->DelXLine(xl);
            xlm->AddXLine(xl);
        }
    } else {
        Anope::string smask, sby, sreason, suid;
        time_t expires;

        data["mask"] >> smask;
        data["by"] >> sby;
        data["reason"] >> sreason;
        data["uid"] >> suid;
        data["expires"] >> expires;

        xl = new XLine(smask, sby, expires, sreason, suid);
        xlm->AddXLine(xl);
    }

    data["created"] >> xl->created;
    xl->manager = xlm;

    return xl;
}

void XLineManager::RegisterXLineManager(XLineManager *xlm) {
    XLineManagers.push_back(xlm);
}

void XLineManager::UnregisterXLineManager(XLineManager *xlm) {
    std::list<XLineManager *>::iterator it = std::find(XLineManagers.begin(),
            XLineManagers.end(), xlm);

    if (it != XLineManagers.end()) {
        XLineManagers.erase(it);
    }
}

void XLineManager::CheckAll(User *u) {
    for (std::list<XLineManager *>::iterator it = XLineManagers.begin(),
            it_end = XLineManagers.end(); it != it_end; ++it) {
        XLineManager *xlm = *it;

        if (xlm->CheckAllXLines(u)) {
            break;
        }
    }
}

Anope::string XLineManager::GenerateUID() {
    Anope::string id;
    int count = 0;
    do {
        id.clear();

        if (++count > 10) {
            Log(LOG_DEBUG) << "Unable to generate XLine UID";
            break;
        }

        for (int i = 0; i < 10; ++i) {
            char c;
            do {
                c = (rand() % 75) + 48;
            } while (!isupper(c) && !isdigit(c));
            id += c;
        }
    } while (XLinesByUID->count(id) > 0);

    return id;
}

XLineManager::XLineManager(Module *creator, const Anope::string &xname,
                           char t) : Service(creator, "XLineManager", xname), type(t), xlines("XLine") {
}

XLineManager::~XLineManager() {
    this->Clear();
}

const char &XLineManager::Type() {
    return this->type;
}

size_t XLineManager::GetCount() const {
    return this->xlines->size();
}

const std::vector<XLine *> &XLineManager::GetList() const {
    return this->xlines;
}

void XLineManager::AddXLine(XLine *x) {
    if (!x->id.empty()) {
        XLinesByUID->insert(std::make_pair(x->id, x));
    }
    this->xlines->push_back(x);
    x->manager = this;
}

void XLineManager::RemoveXLine(XLine *x) {
    /* called from the destructor */

    std::vector<XLine *>::iterator it = std::find(this->xlines->begin(),
                                        this->xlines->end(), x);

    if (!x->id.empty()) {
        std::multimap<Anope::string, XLine *, ci::less>::iterator it2 =
            XLinesByUID->find(x->id), it3 = XLinesByUID->upper_bound(x->id);
        for (; it2 != XLinesByUID->end() && it2 != it3; ++it2)
            if (it2->second == x) {
                XLinesByUID->erase(it2);
                break;
            }
    }

    if (it != this->xlines->end()) {
        this->SendDel(x);
        this->xlines->erase(it);
    }
}

bool XLineManager::DelXLine(XLine *x) {
    std::vector<XLine *>::iterator it = std::find(this->xlines->begin(),
                                        this->xlines->end(), x);

    if (!x->id.empty()) {
        std::multimap<Anope::string, XLine *, ci::less>::iterator it2 =
            XLinesByUID->find(x->id), it3 = XLinesByUID->upper_bound(x->id);
        for (; it2 != XLinesByUID->end() && it2 != it3; ++it2)
            if (it2->second == x) {
                XLinesByUID->erase(it2);
                break;
            }
    }

    if (it != this->xlines->end()) {
        this->SendDel(x);

        x->manager = NULL; // Don't call remove
        delete x;
        this->xlines->erase(it);

        return true;
    }

    return false;
}

XLine* XLineManager::GetEntry(unsigned index) {
    if (index >= this->xlines->size()) {
        return NULL;
    }

    XLine *x = this->xlines->at(index);
    x->QueueUpdate();
    return x;
}

void XLineManager::Clear() {
    std::vector<XLine *> xl;
    this->xlines->swap(xl);

    for (unsigned i = 0; i < xl.size(); ++i) {
        XLine *x = xl[i];
        if (!x->id.empty()) {
            XLinesByUID->erase(x->id);
        }
        delete x;
    }
}

bool XLineManager::CanAdd(CommandSource &source, const Anope::string &mask,
                          time_t expires, const Anope::string &reason) {
    for (unsigned i = this->GetCount(); i > 0; --i) {
        XLine *x = this->GetEntry(i - 1);

        if (x->mask.equals_ci(mask)) {
            if (!x->expires || x->expires >= expires) {
                if (x->reason != reason) {
                    x->reason = reason;
                    source.Reply(_("Reason for %s updated."), x->mask.c_str());
                } else {
                    source.Reply(_("%s already exists."), mask.c_str());
                }
            } else {
                x->expires = expires;
                if (x->reason != reason) {
                    x->reason = reason;
                    source.Reply(_("Expiry and reason updated for %s."), x->mask.c_str());
                } else {
                    source.Reply(_("Expiry for %s updated."), x->mask.c_str());
                }
            }

            return false;
        } else if (Anope::Match(mask, x->mask) && (!x->expires
                   || x->expires >= expires)) {
            source.Reply(_("%s is already covered by %s."), mask.c_str(), x->mask.c_str());
            return false;
        } else if (Anope::Match(x->mask, mask) && (!expires || x->expires <= expires)) {
            source.Reply(_("Removing %s because %s covers it."), x->mask.c_str(),
                         mask.c_str());
            this->DelXLine(x);
        }
    }

    return true;
}

XLine* XLineManager::HasEntry(const Anope::string &mask) {
    std::multimap<Anope::string, XLine *, ci::less>::iterator it =
        XLinesByUID->find(mask);
    if (it != XLinesByUID->end())
        for (std::multimap<Anope::string, XLine *, ci::less>::iterator it2 =
                    XLinesByUID->upper_bound(mask); it != it2; ++it)
            if (it->second->manager == NULL || it->second->manager == this) {
                it->second->QueueUpdate();
                return it->second;
            }
    for (unsigned i = 0, end = this->xlines->size(); i < end; ++i) {
        XLine *x = this->xlines->at(i);

        if (x->mask.equals_ci(mask)) {
            x->QueueUpdate();
            return x;
        }
    }

    return NULL;
}

XLine *XLineManager::CheckAllXLines(User *u) {
    for (unsigned i = this->xlines->size(); i > 0; --i) {
        XLine *x = this->xlines->at(i - 1);

        if (x->expires && x->expires < Anope::CurTime) {
            this->OnExpire(x);
            this->DelXLine(x);
            continue;
        }

        if (this->Check(u, x)) {
            this->OnMatch(u, x);
            return x;
        }
    }

    return NULL;
}

void XLineManager::OnExpire(const XLine *x) {
}
